bin/fetchl10nstrings.js

```diff
@@ -0,0 +1,341 @@
+        fetch(response.url)
+          .then(assureResponse)
+          .then(resolve);
+      }, 2000);
+      break;
+    case 404:
+      msg = warn(`l10n File is 404 : ${response.url}`);
+      console.warn(msg);
+      // Node Response doesn't expose constructor, so fake it till you make it!
+      const fakeResponse = {
+        text() {
+          return Promise.resolve("");
+        }
+      };
+      resolve(fakeResponse);
+      break;
+    default:
+      msg = error(`Could not handle ${response.status} for: ${response.url}`);
+      reject(new Error(msg));
+    }
+  });
+}
+/**
+ * Validate the data against the canonical default object.
+ * Invalid properties are console.log()'ed.
+ *
+ * @param  {Objects[]} results results to be validated.
+ * @param  {String} locale The corresponding locale.
+ * @return {Object[]} the validated (unmodified) results.
+ */
+function validateResults(results, locale) {
+  let tempObj = results.reduce(
+    (prev, next) => Object.assign(prev, next), {}
+  );
+  let missingKeys = Object.keys(defaultLocale)
+    .filter(key => !tempObj.hasOwnProperty(key));
+  if (missingKeys.length) {
+    console.warn(`
+${warn("WARNING:")} Locale "${locale}" is missing keys. The en-US locale will fill the gaps:
+ ${notice("* " + missingKeys.join("\n * "))}
+    `);
+  }
+  return results;
+}
+/**
+ * Writes the resulting JSON to the file system.
+ *
+ * @param  {String} data   The string to write to disk.
+ * @param  {String} locale The locale that this data is for.
+ * @return {Promise} Resolves when writing is done.
+ */
+function writeToDisk(data, locale) {
+  return new Promise((resolve, reject) => {
+    const dir = path.resolve(`${l10nPath}/${locale}/`);
+    if (!fs.existsSync(dir)) {
+      fs.mkdirSync(dir);
+    }
+    fs.writeFile(`${dir}/strings.json`, data, "utf8", (err) => {
+      if (err) {
+        return reject(err);
+      }
+      resolve();
+    });
+  });
+}
+/**
+ * Process DTD files by:
+ *   - Split on each new line,
+ *   - then filters out anything that doesn't start with "<!ENTITY"
+ *   - then removes "<" and ">" and additional quotation marks.
+ *   - then combines the remaining key/value pair into the result object.
+ *
+ * @param  {String} text Text to be processed.
+ * @return {Object} The resulting object with the key value pairs.
+ */
+function processDTD(text) {
+  return text.split("\n")
+    .filter(line => line.trim().startsWith("<!ENTITY"))
+    .map(
+      line => line.replace("<!ENTITY ", "")
+      .replace(">", "")
+      .replace(/\"/g, "")
+      .split(/\s(.+)?/)
+      .filter(item => item)
+    )
+    .map(cleanUpNameValuePairs)
+    .reduce(toObject, {});
+}
+/**
+ * Reduces key value pairs into an Object.
+ *
+ * @param  {Object} obj The object to reduce into.
+ * @param  {String[]} nameValue The name value pair to convert to properties
+ * @return {Object} The resulting object
+ */
+function toObject(obj, nameValue) {
+  obj[nameValue.shift()] = nameValue.shift();
```

Comment:
This works just as well:
```javascript
let [key, value] = nameValue
obj[key] = value;
```
One more line, but clearer.

This is a nit, I don't mind shift in isolation, but i have to many scars from perl in a sprawling code base.